journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegment.java

```diff
@@ -173,23 +188,56 @@ public boolean isOpen() {
   /** Closes the segment. */
   @Override
   public void close() {
-    writer.close();
+    open = false;
     readers.forEach(MappedJournalSegmentReader::close);
     IoUtil.unmap(buffer);
-    open = false;
   }
 
   /** Deletes the segment. */
   public void delete() {
+    open = false;
+    markForDeletion();
+    if (readers.isEmpty()) {
+      safeDelete();
+    }
+  }
+
+  private void safeDelete() {
+    if (!readers.isEmpty()) {
+      throw new JournalException(
+          String.format(
+              "Cannot delete segment file. There are %d readers referring to this segment.",
+              readers.size()));
+    }
     try {
-      Files.deleteIfExists(file.file().toPath());
+      IoUtil.unmap(buffer);
+      Files.deleteIfExists(file.getFileMarkedForDeletion());
     } catch (final IOException e) {
-      throw new JournalException(e);
+      LOG.warn(
+          "Could not delete segment {}. File to delete {}. This can lead to increased disk usage.",
+          this,
+          file.getFileMarkedForDeletion(),
+          e);
     }
   }
 
   @Override
   public String toString() {
     return toStringHelper(this).add("id", id()).add("index", index()).toString();
   }
+
+  private void markForDeletion() {
+    if (markedForDeletion) {
```

Comment:
:question: I think there is still a race condition :smile: But possibly not? This is called from delete, correct? If we guarantee that delete is only called by the same thread, then it's fine - doing the read of `markedForDeletion` and the write later should be OK. If not, then we should do a compare and swap. Not a blocker if this isn't the case.

Reply:
`delete` is always called by the same thread - by the writer. Similar to other `write` methods, it is not enforced with in the journal. But we assume this is enforced in the higher layer. So we can assume as all the methods that can modify journal is called by the raft thread.